src/http/request.cr

```diff
@@ -29,8 +29,12 @@ class HTTP::Request
   # This property is not used by `HTTP::Client`.
   property remote_address : String?
 
-  def initialize(@method : String, @resource : String, headers : Headers? = nil, body : String | Bytes | IO | Nil = nil, @version = "HTTP/1.1")
-    @headers = headers.try(&.dup) || Headers.new
+  def initialize(method : String, resource : String, headers : Headers? = nil, body : String | Bytes | IO | Nil = nil, version = "HTTP/1.1")
```

Comment:
I think it would be better to provide a custom `self.new` instead of `initialize`. The `dup_headers` logic is also unnecessary, it can just be duped right here:

```cr
def self.new(method : String, resource : String, headers : Headers? = nil, body : String | Bytes | IO | Nil = nil, version = "HTTP/1.1")
  # duplicate headers to prevent the request from modifying data that the user might hold.
  new(method, resource, headers.try(&.dup), body, version, internal: nil)
end
  
private def initialize(@method : String, @resource : String, headers : Headers? = nil, body : String | Bytes | IO | Nil = nil, @version = "HTTP/1.1", *, internal)
  @headers ||= Headers.new
  self.body = body
end
```

Reply:
I though about internal but it looks strange to use a named argument like thar. In fact it was my first way of doing it. Maybe we should actually expose the dup_headers argument publicly in case someone needs this performance and they know what they are doing. Thoughts?

Reply:
Yes, it's a bit strange. Alternatively, `headers` could just be differently named in the internal method (for example `safe_headers`) and we call it with a named argument.
In any case, I think the duplication should not happen in the internal method.

I don't think exposing it is necessary. When you create a `Request` without `headers` argument, it won't be duped and you can just use `request.headers` to build the headers.

Reply:
I can't find any of these ways objectively better so for now I'd like to leave it like this.

Reply:
Plus this PR doesn't change any API so we can take care of this little detail later.

Reply:
IMO having the duplication logic in the internal constructor is unnecessary complexity. But it's okay to leave it as is.

Reply:
I changed it (so I could try `fixup`, though I already squashed... but it's useful!)